dbms/src/Storages/S3/S3GCManager.cpp

```diff
@@ -0,0 +1,497 @@
+            for (const auto & prefix : prefixes)
+            {
+                const auto filename_view = S3FilenameView::fromStoreKeyPrefix(prefix.GetPrefix());
+                RUNTIME_CHECK(filename_view.type == S3FilenameType::StorePrefix, prefix.GetPrefix());
+                all_store_ids.emplace_back(filename_view.store_id);
+            }
+            return PageResult{.num_keys = prefixes.size(), .more = true};
+        });
+
+    return all_store_ids;
+}
+
+std::unordered_set<String> S3GCManager::getValidLocksFromManifest(const String & manifest_key)
+{
+    // download the latest manifest from S3 to local file
+    const String local_manifest_path = getTemporaryDownloadFile(manifest_key);
+    downloadFile(*client, client->bucket(), local_manifest_path, manifest_key);
+    LOG_INFO(log, "Download manifest, from={} to={}", manifest_key, local_manifest_path);
+
+    // parse lock from manifest
+    PS::V3::CheckpointProto::StringsInternMap strings_cache;
+    using ManifestReader = DB::PS::V3::CPManifestFileReader;
+    auto reader = ManifestReader::create(ManifestReader::Options{.file_path = local_manifest_path});
+    auto mf_prefix = reader->readPrefix();
+
+    while (true)
+    {
+        // TODO: calculate the valid size of each CheckpointDataFile in the manifest
+        auto part_edit = reader->readEdits(strings_cache);
+        if (!part_edit)
+            break;
+    }
+
+    std::unordered_set<String> locks;
+    while (true)
+    {
+        auto part_locks = reader->readLocks();
+        if (!part_locks)
+            break;
+        locks.merge(part_locks.value());
+    }
+
+    return locks;
+}
+
+void S3GCManager::removeOutdatedManifest(const CheckpointManifestS3Set & manifests, const Aws::Utils::DateTime * const timepoint)
+{
+    if (timepoint == nullptr)
+    {
+        for (const auto & mf : manifests.objects())
+        {
+            // store tombstoned, remove all manifests
+            deleteObject(*client, client->bucket(), mf.second.key);
+            LOG_INFO(
+                log,
+                "remove outdated manifest because of store tombstone, key={} mtime={}",
+                mf.second.key,
+                mf.second.last_modification.ToGmtString(Aws::Utils::DateFormat::ISO_8601));
+        }
+        return;
+    }
+
+    assert(timepoint != nullptr);
+    // clean the outdated manifest files
+    const auto outdated_mfs = manifests.outdatedObjects(config.manifest_preserve_count, config.manifest_expired_hour, *timepoint);
+    for (const auto & mf : outdated_mfs)
+    {
+        // expired manifest, remove
+        deleteObject(*client, client->bucket(), mf.key);
+        LOG_INFO(
+            log,
+            "remove outdated manifest, key={} mtime={}",
+            mf.key,
+            mf.last_modification.ToGmtString(Aws::Utils::DateFormat::ISO_8601));
+    }
+}
+
+String S3GCManager::getTemporaryDownloadFile(String s3_key)
```

Comment:
```suggestion
String S3GCManager::getTemporaryDownloadFile(const String & s3_key)
```

Reply:
this can not be `const &` because I replace '/' with '_' in the s3_key for a download name